napari/layers/_shapes_layer/shape_list.py

```diff
@@ -0,0 +1,616 @@
+    _mesh_vertices_offsets : np.ndarray
+         Qx2 array of offsets of vertices of triangles for shapes. For vertices
+         corresponding to faces these are 0. For vertices corresponding to
+         edges these values should be scaled and added to the
+         `_mesh_vertices_centers` to get the actual vertex positions.
+         The scaling corresponds to the width of the edge
+    _mesh_vertices_index : np.ndarray
+         Qx2 array of the index (0, ..., N-1) of each shape that each vertex
+         corresponds and the mesh type (0, 1) for face or edge.
+    _mesh_triangles : np.ndarray
+        Px3 array of vertex indices that form the mesh triangles
+    _mesh_triangles_index : np.ndarray
+        Px2 array of  the index (0, ..., N-1) of each shape that each triangle
+        corresponds and the mesh type (0, 1) for face or edge.
+    _mesh_triangles_colors : np.ndarray
+        Px4 array of the rgba color of each triangle
+    _mesh_triangles_z_order : np.ndarray
+        Length P array of the z order of each triangle. Must be a permutation
+        of (0, ..., P-1)
+    _shape_types : dict
+        Dictionary of supported shape types and their corresponding objects.
+    """
+    _mesh_types = ['face', 'edge']
+
+    _shape_types = ({'rectangle': Rectangle, 'ellipse': Ellipse, 'line': Line,
+                    'path': Path, 'polygon': Polygon})
+
+    def __init__(self, data, *, shape_type='rectangle', edge_width=1,
+                 edge_color='black', face_color='white', opacity=1, z_index=0):
+
+        self.shapes = []
+        self._vertices = np.empty((0, 2))
+        self._index = np.empty((0), dtype=int)
+        self._z_index = np.empty((0), dtype=int)
+        self._z_order = np.empty((0), dtype=int)
+
+        self._mesh_vertices = np.empty((0, 2))
+        self._mesh_vertices_centers = np.empty((0, 2))
+        self._mesh_vertices_offsets = np.empty((0, 2))
+        self._mesh_vertices_index = np.empty((0, 2), dtype=int)
+        self._mesh_triangles = np.empty((0, 3), dtype=np.uint32)
+        self._mesh_triangles_index = np.empty((0, 2), dtype=int)
+        self._mesh_triangles_colors = np.empty((0, 4))
+        self._mesh_triangles_z_order = np.empty((0), dtype=int)
+
+        if issubclass(type(data), Shape):
+            # If a single shape has been passed
+            self.add(data)
+        elif len(data) > 0:
+            if issubclass(type(data[0]), Shape):
+                # If list of shapes has been passed
+                for d in data:
+                    self.add(d)
+            elif np.array(data[0]).ndim == 1:
+                # If a single array for a shape has been passed
+                self.add(data, shape_type=shape_type, edge_width=edge_width,
+                         edge_color=edge_color, face_color=face_color,
+                         opacity=opacity, z_index=z_index)
+            else:
+                # If list of arrays has been passed
+                for i, d in enumerate(data):
```

Comment:
I'll argue again here to not have this option and instead just make a user friendly way to create Shape objects. I just think that this will lead to a lot of parsing errors, hard to validate input and ways for people to get their shape list into a error state they can't get out of.

Reply:
I have pulled this out of the `shape_list` which is now cleaner and put it into the `model` file. I hope this is a step in the right direction, though maybe it is not enough!